### PR TITLE
Add Universe Opt In Source

### PR DESCRIPTION
**Pull request by localjo**
I'd like to add support for a Universe API parameter to subscribe-email that allows the source of an opt-in to be defined

https://github.com/SparkartGroupInc/sparkart-services/blob/master/doc/api/contacts.md

`contact[lists_opt_in_source]`

### CODE REVIEW

**Comment by josiahsprague** [comment]
Something about adding arbitrary parameters for only one API doesn't settle right with me. I'm trying to think if there is a good way to wrap service-specific options so that it's clear that an option only works with service x.

**Comment by localjo** [comment]
Thinking through some ways to approach this while being conscious of how other mailing list providers handle things so we don't end up with an unwieldy set of options that are different for each provider. What if we did something like this?

```
var mySubscribeForm = new SubscribeEmail({
  element: '#subscribe-form',
  service: 'universe',
  key: 'api-key-here',
  customFields: {
    'contact[lists_opt_in_source]': {
      type: 'hidden',
      value: 'Some string'
    },
    'contact[first_name]': {
      type: 'text',
      label: 'First Name'
    }
  }
});
```

That way we could account for any possible fields someone might want to add to their subscribe form. In Universe, the field options are pretty consistent. But in MailChimp and SendGrid, custom fields can be added for whatever an account manager might want to collect.

**Comment by localjo** [comment]
Another way of looking at it is that this could be a good use case for custom templates. Just add a custom template with `<input type="hidden" name="contact[lists_opt_in_source]" value="Opt In Source Name">`. I actually like this option. But maybe writing some verbose documentation on how to achieve it would be useful. Thoughts @pushred ?

**Comment by pushred** [comment]
Yeah you're right, custom fields is exactly how [MailChimp](http://blog.mailchimp.com/tracking-signup-pages-via-the-mailchimp-embedded-forms/) and [Campaign Monitor](http://help.campaignmonitor.com/topic.aspx?t=185) handle this for instance. Universe also supports custom fields but this standard parameter is better because we're tracking email imports using it as well.

There's some advantage to the `customFields` approach, in the case of this particular parameter it'd likely be populated from JS anyway — site domain, a path, page title. It's more direct to fetch that data at instantiation rather than after the template has been rendered. We have a JS template at our disposal, which is really just an orderly way of doing DOM manipulation at the end of the day. This also allows the default template to be used more often.

I think what I was after here was a way of baking this in such that the source is guaranteed to be submitted, since it's important for compliance and list hygiene. But if other services generally require additional setup anyway then I agree some documentation would be helpful.

**Comment by josiahsprague** [comment]
How about something like;
```
hiddenFields: [
  name: 'contact[lists_opt_in_source]',
  value: 'Some string'
]
```

For adding data that should be submitted by the form but doesn't affect what's displayed on a page. I think it makes sense to add non-displayed elements, such as hidden inputs, without making a custom template, but if displayed elements need to be added, such as text inputs, a custom template should be used.

**Comment by localjo** [comment]
Added an option to do;

```
hiddenFields: [
  {
    name: 'contact[lists_opt_in_source]',
    value: 'My Marketing Campaign'
  }
]
```

This can also be used for other hidden inputs in order to add data that should be submitted by the form but doesn't affect what's displayed on a page. I think it makes sense to add non-displayed elements, such as hidden inputs, without making a custom template, but if displayed elements need to be added, such as text inputs, a custom template should be used.

Also added tests and updated the README.

File: README.md [hunk 1]
```diff
@@ -62,6 +62,18 @@ var mySubscribeForm = new SubscribeEmail({
 ### `url`
 **(Required for MailChimp)** A string of the `<form action="">` attribute generated by MailChimp which contains MailChimp authentication information. You can get this from MailChimp under "Signup forms > Embedded form code > Naked" and copying just the value from the `<form action="">` attribute. It should follow this format: `http://{username}.{data center}.list-manage.com/subscribe/post?u={user id}&id={list id}`.
 
+### `hiddenFields`
+An optional array of objects used to add `<input type="hidden">` elements to the default template to allow additional data to be submitted along with a form. As an example, on Universe forms this can be used to track an opt-in source using the [`lists_opt_in_source` parameter](https://github.com/SparkartGroupInc/sparkart-services/blob/master/doc/api/contacts.md);
```
Comment (pushred) [inline README.md]: Unfortunately can't link to the Universe docs at the moment, that's a private repo. Maybe reference the MailChimp example instead?

File: src/subscribe-form.hbs [hunk 2]
```diff
@@ -1,4 +1,7 @@
 <form id="{{id}}" data-subscribe-instance="{{instanceId}}" class="{{namespace}}">
+{{#each hiddenFields}}
+<input class="{{../namespace}}__input {{../namespace}}__input--hidden" type="hidden" name="{{name}}" value="{{value}}">
```
Comment (pushred) [inline src/subscribe-form.hbs]: What's the value of CSS classes on hidden inputs?

File: test/tests.js [hunk 3]
```diff
@@ -154,4 +154,28 @@ describe('Subscribe Email Module', function() {
...
+          {
+            name: 'testField2',
+            value: 'testValue2'
+          }
+        ]
+      });
+      var testField1 = $('#test-element .subscribe-email__input--hidden[name="testField1"]').val('testValue1');
+      assert(testField1.length > 0);
+      var testField2 = $('#test-element .subscribe-email__input--hidden[name="testField2"]').val('testValue2');
+      assert(testField2.length > 0);
```
Comment (pushred) [inline test/tests.js]: Confused about what this is testing.. why are you setting values on the fields? Shouldn't the assertion be checking for the values from the config?

File: src/subscribe-form.hbs (hunk 2 above)
Comment (localjo) [inline src/subscribe-form.hbs]: Well, in theory those classes could be used for JS too. I was using them as handles for the tests, though that's not really necessary. They're not really necessary here, just added for consistency with the rest of the elements.

**Comment by localjo** [comment]
@pushred I'm not sure if this is still needed, but I think it's ready to merge.

**Comment by localjo** [comment]
Doing some housekeeping and closing this PR due to lack of activity. 
giphy

**Pull request closed without merging**